File: dsa/19-Graphs/GeeksForGeeks/kruskal.java
// User function Template for Java

import java.util.Arrays;

class Solution {
    static int kruskalsMST(int V, int[][] edges) {
        dsu uf = new dsu();
        uf.init(V);
        Arrays.sort(edges, (a, b) -> a[2] - b[2]); // O(ElogE)
        int finalCost = 0;
        int cnt = 0;

        for (int i = 0; cnt < V - 1; i++) { // can do i<edges.length but code will run a bit more then

            // O(V)
            int[] edge = edges[i];
            int parA = uf.find(edge[0]);
            int parB = uf.find(edge[1]);
            if (parA != parB) {
                uf.union(parA, parB);
                finalCost += edge[2];
                cnt++;
            }
        }

        return finalCost;
    }
}

class dsu {
    int n;
    int par[];
    int rank[];

    public void init(int V) {
        n = V;
        par = new int[n];
        rank = new int[n];
        for (int i = 0; i < n; i++) {
            par[i] = i;
        }
    }

    public int find(int x) {
        if (x == par[x]) {
            return x;
        }

        return par[x] = find(par[x]);
    }

    public void union(int x, int y) {
        int px = find(x);
        int py = find(y);

        if (rank[px] == rank[py]) {
            par[py] = px;
            rank[px]++;
        } else if (rank[px] < rank[py]) {
            par[px] = py;
        } else {
            par[py] = px;
        }
    }
}

File: dsa/19-Graphs/GeeksForGeeks/kruskal.cpp
#include <bits/stdc++.h>
using namespace std;

class DSU {
   private:
    int n;
    vector<int> par;
    vector<int> rank;

   public:
    void init(int V) {
        this->n = V;
        this->rank.resize(n, 0);
        for (int i = 0; i < n; i++) {
            par.push_back(i);
        }
    }

    int find(int x) {
        if (x == par[x]) {
            return x;
        }

        return par[x] = find(par[x]);
    }

    void unionSet(int x, int y) {
        int px = find(x);
        int py = find(y);

        if (rank[px] == rank[py]) {
            par[py] = px;
            rank[px]++;
        } else if (rank[px] < rank[py]) {
            par[px] = py;
        } else {
            par[py] = px;
        }
    }
};

class Solution {
   public:
    int kruskalsMST(int V, vector<vector<int>>& edges) {
        DSU uf;
        uf.init(V);
        sort(edges.begin(), edges.end(),
             [](const vector<int>& x, const vector<int>& y) {
                 return x[2] < y[2];
             });
        int cost = 0;
        int cnt = 0;
        for (int i = 0; cnt < V - 1; i++) {
            vector<int> edge = edges[i];
            int px = uf.find(edge[0]);
            int py = uf.find(edge[1]);

            if (px != py) {
                uf.unionSet(px, py);
                cost += edge[2];
                cnt++;
            }
        }

        return cost;
    }
};